pkg/synchronizer/opstate.go

```diff
@@ -0,0 +1,198 @@
+// SPDX-FileCopyrightText: 2022-present Intel Corporation
+// SPDX-FileCopyrightText: 2020-present Open Networking Foundation <info@opennetworking.org>
+//
+// SPDX-License-Identifier: Apache-2.0
+
+// Package synchronizer implements a synchronizer for Aether models
+package synchronizer
+
+import (
+	"github.com/onosproject/sdcore-adapter/pkg/gnmi"
+	"github.com/onosproject/sdcore-adapter/pkg/metrics"
+	"strconv"
+	"time"
+)
+
+// Information we keep track of when we call prometheus. The endpoint of the
+// prometheus server, as well as the enterprise and site that the opstate will
+// be pushed to.
+type prometheusInfo struct {
+	entID    string
+	siteID   string
+	endpoint string
+}
+
+// given an IMSI, return the simCard that has that IMSI, or nil if none exists
+func (s *Synchronizer) getSimCardByImsi(site *Site, imsi uint64) *SimCard {
+	for _, sim := range site.SimCard {
+		if sim.Imsi != nil && *sim.Imsi == imsi {
+			return sim
+		}
+	}
+	return nil
+}
+
+// given an IMSI, return the device that has a simcart that has that IMSI, or nil if none exists
+func (s *Synchronizer) getDeviceByImsi(site *Site, imsi uint64) *Device {
+	sim := s.getSimCardByImsi(site, imsi)
+	if sim == nil {
+		return nil
+	}
+	for _, dev := range site.Device {
+		if *dev.SimCard == *sim.SimId {
+			return dev
+		}
+	}
+	return nil
+}
+
+// get a prometheus metrics fetcher, caching previously-created fetchers
+func (s *Synchronizer) getFetcher(endpoint string) (*metrics.Fetcher, error) {
+	mf, okay := s.prometheus[endpoint]
+	if okay {
+		return mf, nil
+	}
+
+	mf, err := metrics.NewFetcher(endpoint)
+	if err != nil {
+		return nil, err
+	}
+	s.prometheus[endpoint] = mf
+	return mf, nil
+}
+
+// update Device.IPAddress opstate
+func (s *Synchronizer) updateDeviceIPAddress(config *gnmi.ConfigForest, entID string, siteID string, imsi uint64, mobileIP string, connected float64) {
+	// Hold the lock while updating the config tree
+	config.Mu.Lock()
+	defer config.Mu.Unlock()
+
+	entRoot, okay := config.Configs[entID]
+	if !okay {
+		// Rare corner case -- the enterprise has disappeared. Maybe some other thread deleted it
+		log.Warnf("Failed to get enterprise %s", entID)
+		return
+	}
+	enterprise := entRoot.(*RootDevice)
+
+	site, okay := enterprise.Site[siteID]
+	if !okay {
+		// Rare corner case -- the site has disappeared. Maybe some other thread deleted it
+		log.Warnf("Failed to get site %s", siteID)
+		return
+	}
+
+	device := s.getDeviceByImsi(site, imsi)
+	if device == nil {
+		// Can happen if the device is reported in prometheus, but does not exist in
+		// ROC. Rare in real life, but plausible in test infrastructure.
+		// Message is currently disabled, as it is fairly noisy.
+		//log.Warnf("Failed to find device %v in site %s of enterprise %s", imsi, siteID, entID)
```

Comment:
probably should be `Debugf` anyway

Reply:
I wish we had a level that was even further down than debug.